### DISCUSSION TITLE
periodic_dirac_kernel

### DISCUSSION THREAD

**Discussion by Ali1990dashti** [idaholab/moose#20055] [Q&A General] [answered]
Dear community,

I am trying to make a heat storage simulation using a MOOSE based application. In my model I have a hot and a cold well. In both wells I inject for six month and then produce for the next six month. For time stepping I am using `FunctionDT` functionality of MOOSE. For example the follwing lines represente my time steps (small time steps when the injection or production starts):
```
time_t = '0 864000 15552000 15552001 16416000 31104000 31104001 31968000 46656000 46656001 47520000 62208000'
time_dt= '60 172800 172800 60 172800 172800 60 172800 172800 60 172800 172800'
```

My main issue is that in wells my Dirac kernel is fluctiating: 6 month injection and 6 month production. In other words from time 0 to 15552000 (6 month) the flow on the point is for example -2 (injection), whereas from 15552001 to 31104000 (12 month) it is +2 (production). Now, I am wondering how I can make this shift in my dirac kernel. How can I define such a Dirac kernel in MOOSE that follows my times?
In advance I very much appreciate any help.

**Comment by avtarsinghh1991** [comment]
Look at the following, it may be helpful.

https://mooseframework.inl.gov/syntax/Executioner/TimeStepper/index.html

Best
Avtar

**Reply by Ali1990dashti** [reply]
Dear @asingh-mit,

Thanks for devoting time to my issue. 
I read to documentation but still I did not find a way to make the dirac kernel periodic. Do you know any sample doing such a thing?

Thanks for your help
Ali

**Reply by avtarsinghh1991** [reply]
Dear @Ali1990dashti 

As far I understand, you want to apply periodic boundary conditions. Am I right?

If so you can generate your functions for doing so. I hope you find the following links helpful.

https://mooseframework.inl.gov/syntax/Functions/index.html

https://mooseframework.inl.gov/source/actions/AddPeriodicBCAction.html


Best
Avtar

**Reply by Ali1990dashti** [reply]
Dear @asingh-mit 

Thanks fror your feedback. I will go through the links your provided.

Cheers,
Ali

**Comment by GiudGiud** [comment, marked as answer]
Hello

Can you use the FunctionDiracSource?
https://mooseframework.inl.gov/source/dirackernels/FunctionDiracSource.html

This will allow you to specify when the dirac source is 0 or non 0, as shown in the example with a ParsedFunction

Guillaume

**Comment by WilkAndy** [comment]
If you're using a [PorousFlowSink](https://mooseframework.inl.gov/source/bcs/PorousFlowSink.html) or one of its derived classes (PiecewiseLinear, etc) then you can use the `flux_function` input to specify what you want.

**Reply by WilkAndy** [reply]
Or, if you're using a [PorousFlowPolyLineSink](https://mooseframework.inl.gov/source/dirackernels/PorousFlowPolyLineSink.html) then the `multiplying_var` can be made periodic as you desire.

**Comment by cpgr** [comment]
Yet another option is to use the `Controls` system to turn the Dirac kernels on off - so you can have an injection point source and a production point source, and toggle them in the input file - see https://github.com/idaholab/moose/blob/next/modules/porous_flow/examples/ates/ates.i for an example.

**Reply by cpgr** [reply]
Documentation of above example is also available - https://mooseframework.inl.gov/modules/porous_flow/ates.html
